[Nicole Lee Siying] iP

nickeltea opened the pull request:
## **Duke**
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke is your friendly task-manager. It can:
- Track **Todo**s
- Track **Event**s
- Track **Deadline**s
- Mark Tasks as complete or incomplete.
- Remove Tasks.

Duke is:
1. Test-based
2. ~~Easy~~ SUPER easy to learn!

Features:
- [x] Managing Tasks 👍 
- [ ] Reminders (coming soon)

If you're Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
        new Duke("data/duke.txt").run(); 
}
```

Commit:
Add Gradle support

nickeltea committed:
Level-1

nickeltea committed:
Level-2

nickeltea committed:
Level-3

nickeltea committed:
Level-4

nickeltea committed:
A-TextUiTesting

nickeltea committed:
Level-5

nickeltea committed:
Level-6

nickeltea committed:
Added. documentation.

nickeltea committed:
Save tasks in the hard disk.

The task list resets every time the program is run.

Tasks added to the task list are not saved. The task is is blank whenever the program is started.

Let's save the tasks in the hard disk automatically whenever the task list changes and load the data from the hard disk when Duke starts up.

nickeltea committed:
Make Duke understand dates and times

Duke accepts any String as a deadline and treats dates/times given as Strings.

Some Strings are not valid dates and times.

Let's teach Duke to accept Strings in a specific date and time format only, and treat this String as a LocalDateTime.

nickeltea committed:
Merge branch 'branch-Level-8'

nickeltea committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

nickeltea committed:
Make the code more OOP

The code can be broken up into classes of related code.

The code is currently not using OOP as much it should.

Let's make the code more OOP by refactoring the code to extract out closely related code as classes.

nickeltea committed:
Divide classes into packages

Classes are currently not in packages.

The classes should be in packages.

Divide all the classes into packages.

nickeltea committed:
Add JUnit tests

Currently no JUnit tests are being used to test this code.

There should be JUnit tests used to test this code.

Let's add JUnit tests to test this code.

nickeltea committed:
Package the App as a JAR file

The app has not been packaged before. After packaging, a MANIFEST.MF file was created.

The app should be packaged as a JAR File. The MANIFEST.MF File should not be version controlled.

Let's package the app as a JAR File.

nickeltea committed:
Add Increments as parallel branches

Currently, the program lacks JavaDoc comments, and does not have a way to find specific tasks.

The prgram should have JavaDoc comments and a way to find tasks.

Let's add JavaDoc comments, tweak the code to comply with a coding standard and give users a way to find a task by searching for a keyword.

nickeltea committed:
Revert "Add Increments as parallel branches"

This reverts commit 5ef461b214ee874e29851aa0efa48883639f83d9.

nickeltea committed:
Merge branch 'branch-Level-9'

nickeltea committed:
Tweak the code to comply with a coding standard

The code currently does not comply with the Java coding standard.

The code should comply with the Java coding standard.

Let's tweak the code to comply with a coding standard.

nickeltea committed:
Add JavaDoc comments

Currently there are not enough JavaDoc comments.

There should be more JavaDoc comments.

Let's add JavaDoc comments.

nickeltea committed:
Tweak the code to comply with a coding standard

Imported classes should always be listed explicitly.

The code should comply with the Java coding standard.

Let's tweak the code to comply with a coding standard.

nickeltea committed:
Stop tracking duke.txt file

duke.txt is used for testing purposes.

duke.txt should be version controlled.

Let's stop version controlling duke.txt.

nickeltea committed:
Revert "Revert "Add Increments as parallel branches""

This reverts commit 9b5895f38c374daa39743bbdcc7e339449f0d709.

nickeltea committed:
Give users a way to find a task by searching for a keyword.

Currently users cannot search for specific tasks.

Users should be able to search for specific tasks.

Let's allow users to search for tasks.

# Conflicts:
#	src/main/java/seedu/duke/Duke.java

nickeltea committed:
Fix error

There is a line in the main method that should not be included.

This line is not necessary.

Let's remove this line.

LiuFangrui commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,33 @@
+package seedu.duke;
+
+public class Duke {
+    private static Storage save;
```
Variable name of type storage could be a noun instead of a verb save

LiuFangrui commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,33 @@
+     * Prints out greeting, initialises the TaskList and opens the scanner to start
+     * receiving commands.
+     */
+    public static void run() {
+        ui.introduction();
+    }
+
+    public static void main(String[] args) {
+        new Duke("data/duke.txt").run();
+
+    }
```
great short and concise beginning! Easy to follow

eesung00 replied:
> great short and concise beginning! Easy to follow

Agree that it is short and concise. But perhaps some of the abstraction are too "deep"? (e.g. command can separate into classes and show how does Duke run). I find it hard to understand with only ui.introduction( ) in the run( ) method in the beginning 😖

LiuFangrui commented on src/main/java/seedu/duke/Event.java:
```diff
@@ -0,0 +1,27 @@
+package seedu.duke;
+public class Event extends Task {
+    protected String at;
```
```suggestion
    protected String date;
```

LiuFangrui commented on src/main/java/seedu/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * Calls the Scanner to wait for the next line of user input.
+     */
+    public void nextCommand() {
+        String command = scan.nextLine();
+        this.parse(command);
+        nextCommand();
+    }
```
extracting this part improved readability

LiuFangrui commented on src/main/java/seedu/duke/Task.java:
```diff
@@ -0,0 +1,73 @@
+     *
+     * @param command The command that determines whether this task is done or not done.
+     */
+    public void toggleDone(String command) {
+        if (command.startsWith("mark ")) {
+            this.isDone = true;
+            System.out.println("Nice! I've marked this task as done: \n" + this.toString());
+        } else if (command.startsWith("unmark ")) {
+            this.isDone = false;
+            System.out.println("OK, I've marked this task as not done yet: \n" + this.toString());
+        }
+    }
```
if the code judges based on the argument command, is it still a toggle?

LiuFangrui commented on src/main/java/seedu/duke/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+
+public class TaskList {
+    private static ArrayList<Task> storage = new ArrayList<>();
+    private Storage saveFile;
```
naming is inconsistant and sometimes confusing across classes

LiuFangrui reviewed (commented):
Overall, your code was easy to read and abstraction was well done. However, naming convention is sometimes inconsistent among classes.

eesung00 commented on src/main/java/seedu/duke/Ui.java:
```diff
@@ -0,0 +1,152 @@
+     * @param deleted The deleted task.
+     * @param size The number of tasks remaining.
+     */
+    public static void deleteText(String deleted, int size) {
```
Will it be more comprehensive if the deleted variable naming change to this? Similar to some other methods variable naming
```suggestion
    public static void deleteText(String deletedTasks, int size) {
```

eesung00 commented on src/main/java/seedu/duke/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
+    protected String by;
```
Variable naming can be better if added some time related info? Similar to Event class as mentioned by @LiuFangrui 
```suggestion
    protected String deadlineTime;
```

eesung00 commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.Iterator;
+import java.util.Scanner;
+
+public class Storage {
```
I like how you have many methods that abstracts the file reading and writing purposes!

eesung00 reviewed (commented):
Overall LGTM, just can make better code structure to improve readability in my opinion.

nickeltea committed:
Add a GUI.

Duke currently does not have a GUI.

Duke should make use of JavaFX to implement a GUI.

Let's add a GUI to Duke.

nickeltea committed:
Merge branch 'branch-Level-10'

nickeltea committed:
Fix JavaDocs.

The JavaDocs in Duke currently do not follow a coding standard.

The JavaDocs in Duke should follow a coding standard.

Let's edit the JavaDocs in Duke to make them follow a coding standard.

nickeltea committed:
Use assert feature to document important assumptions

The code does not use any assertions to help check that the correctness
of the code.

Using assertions helps to check the correctness of the code and ensure
that the correct assumptions are being made at various points in the
code.

As a step towards improving code quality and ensuring the correctness of
the code, let's use assertions to document important assumptions that
should hold at various points in the code.

nickeltea committed:
Improve code quality

The code quality has room for improvement.

Making a few changes would improve the code quality.

Let's make the required changes to improve code quality.

nickeltea committed:
Add FixedDurationTasks

Duke currently does not support tasks that take a fixed amount of time
but do not have a fixed start/end time.

It would be useful for Duke to manage such tasks as well.

Let's make Duke handle FixedDurationTasks.

nickeltea committed:
Merge pull request #1 from nickeltea/branch-A-Assertions

Use assert feature to document important assumptions

nickeltea committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/seedu/duke/Task.java
#	src/main/java/seedu/duke/TaskList.java

nickeltea committed:
Merge pull request #2 from nickeltea/branch-A-CodeQuality

Improve code quality

nickeltea committed:
Merge branch 'branch-BCD-Extension'

# Conflicts:
#	src/main/java/seedu/duke/Storage.java
#	src/main/java/seedu/duke/TaskList.java

nickeltea committed:
Add a User Guide.

Duke currently does not have a user guide.

A user guide would be useful for users.

Let's add a user guide for Duke.

nickeltea committed:
Do some bug fixing.

There are some minor bugs in the code.

We should fix these bugs.

Let's fix the bugs in the code.

nickeltea committed:
Update the user guide

The user guide is inaccurate.

The user guide should be accurate.

Let's update the user guide.

nickeltea committed:
Clean up code.

There are unused lines of code.

There should not be any unused lines of code to improve code quality.

Let's improve code quality by removing the unused lines.
